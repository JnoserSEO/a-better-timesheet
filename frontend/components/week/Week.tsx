import React from 'react'
import Day from '../day/Day'
import Hours from '../day/Hours'
import AddTd from '../day/AddTd'
import { ReducerAction } from '../timesheet'
import weekStyleFilter from '../../utility/week-style-filter'
import Timesheetview from '../../models/view.model'
import FullWeekTuple from '../../models/fullweek.model'
import '../../styles/week/week.scss'

const calculateWeekTotalHours = (days:FullWeekTuple):number => days.map((day)=> day.data.reduce((acc,curr)=>acc+curr.hours,0)).reduce((acc,curr)=>acc+curr,0)

const WeekTotalHours:React.FC<{days:FullWeekTuple}> = ({days})=>{
    const totalHours = calculateWeekTotalHours(days)
    return <div className="week-total-div">
        <div className="week-total-hours"><span>{`You worked a total of `}</span><strong>{`${totalHours} hour${totalHours === 1 ? ' ' : 's'}`}</strong><span>{` this week.`}</span></div>
        <div className="triangle-topleft"></div>
    </div>
}

interface WeekProps {
    days: FullWeekTuple;
    view: Timesheetview;
    dispatch: React.Dispatch<ReducerAction>
}

const Week: React.FC<WeekProps> = ({days,view,dispatch})=>{
    return <div>
        <table className="week">
            <tbody>
                <tr>
                {days.map((day,index)=> weekStyleFilter(view,index,<Day key={index} date={day.date} entries={day.data} /> ))}
                </tr>
                <tr>{days.map((day, index) => weekStyleFilter(view, index, <AddTd key={index} index={index} day={day} dispatch={dispatch} />) )}</tr>
                <Hours days={days} view={view}/>
            </tbody>
        </table>
        <WeekTotalHours days={days} />
    </div>    
}
export default Week