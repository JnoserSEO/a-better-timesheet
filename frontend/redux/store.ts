import { createStore, combineReducers } from 'redux'
import newFullWeek from '../utility/newFullWeek'
import { DayofWeek } from '../models/fullweek.model'
import { EntryProps } from '../components/entry/Entry'

const THIS_WEEK = newFullWeek(new Date())

const weekReducer = (state = THIS_WEEK, action: { type:string,dayOfWeek:DayofWeek,entry?:EntryProps,entryIndex?:number}) => {
    switch (action.type) {
        case 'ADD_ENTRY':
            return {...state,[action.dayOfWeek]:{...state[action.dayOfWeek],data:[...state[action.dayOfWeek].data,action.entry]}};
        case 'UPDATE_ENTRY':
            return { ...state, 
                [action.dayOfWeek]: { ...state[action.dayOfWeek],
                    data:[...state[action.dayOfWeek].data.filter((_,index)=> index !== action.entryIndex ),action.entry]
                }
            };
        case 'REMOVE_ENTRY':
            return {...state,
                [action.dayOfWeek]: {
                    ...state[action.dayOfWeek],
                    data: [...state[action.dayOfWeek].data.filter((_, index)=> index !== action.entryIndex )]
                }
            }; 
        default:
            return state;
    }
}

const reducer = combineReducers({ week: weekReducer})
const store = createStore(reducer)

// store.subscribe(() => {
//     const state = store.getState();
//     const viscolumns = getVisibleColumns(state.columns)
//     console.log(viscolumns)
// })

// store.dispatch(setColumn({ name:'Name' }))
// store.dispatch(setColumn({ name: 'St' }))
// store.dispatch(setColumn({ name: 'Name' }))

export default store