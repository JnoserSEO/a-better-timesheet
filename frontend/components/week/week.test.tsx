import React from 'react'
import {  mount } from 'enzyme'
import store from '../../redux/store'
import { Provider } from 'react-redux'
import Week from './Week'
import { FullWeekTuple } from '../../models/fullweek.model'

beforeEach(() => {
    beforeEach(() => {
        Date.now = jest.fn(() => 1601401562455); // 2020-9-29
    });
});

const days: FullWeekTuple = [{ dayOfWeek: 'Sunday', date: new Date('2020-09-27T00:00:00.000Z'), data: [] },
{ dayOfWeek: 'Monday', date: new Date('2020-09-28T00:00:00.000Z'), data: [{ "hours": 4, "name": "Comp B" }] },
{ dayOfWeek: 'Tuesday', date: new Date('2020-09-29T00:00:00.000Z'), data: [{ "hours": 1, "name": "Comp A" }] },
{ dayOfWeek: 'Wednesday', date: new Date('2020-09-30T00:00:00.000Z'), data: [] },
{ dayOfWeek: 'Thursday', date: new Date('2020-10-01T00:00:00.000Z'), data: [] },
{ dayOfWeek: 'Friday', date: new Date('2020-10-02T00:00:00.000Z'), data: [] },
{ dayOfWeek: 'Saturday', date: new Date('2020-10-03T00:00:00.000Z'), data: [] }]

test('should render Week with two timesheet entries', () => {
    const wrapper = mount(<Provider store={store}><Week days={days} view={'5 day'} dispatch={(x) => x} /></Provider>)
    expect(wrapper.find('.timesheet-entry')).toHaveLength(2)
    expect(wrapper.find('.day-total')).toHaveLength(5)
    expect(wrapper.find('.day-info')).toHaveLength(5)
    expect(wrapper.find('.add-td')).toHaveLength(5)
    expect(wrapper).toMatchSnapshot()
})