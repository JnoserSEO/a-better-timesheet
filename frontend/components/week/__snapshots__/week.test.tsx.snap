// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Week with two timesheet entries 1`] = `ReactWrapper {}`;
